rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRAParticipant.java

```diff
@@ -0,0 +1,249 @@
+            }
+        } else if (result instanceof Response && ((Response) result).getStatus() == Response.Status.ACCEPTED.getStatusCode()) {
+            return true;
+        }
+        
+        return false;
+    }
+
+    Response forget(URI lraId, URI parentId) {
+        return invokeParticipantMethod(forgetMethod, lraId, parentId, FORGET);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Response invokeParticipantMethod(Method method, URI lraId,
+                                             URI parentId, String type) {
+        Object participant = CDI.current().select(id).get();
+        Object result = null;
+
+        switch (method.getParameterCount()) {
+            case 0:
+                result = invokeMethod(type, method, participant);
+                break;
+            case 1:
+                result = invokeMethod(type, method, participant, lraId);
+                break;
+            case 2:
+                result = invokeMethod(type, method, participant, lraId, parentId);
+                break;
+            default:
+                throw new InvalidLRAParticipantDefinitionException(
+                    method.toGenericString() + ": invalid number of arguments: " + method.getParameterCount());
+        }
+
+        return processResult(result, lraId, method, type);
+    }
+
+    private Object invokeMethod(String type, Method method, Object o, Object... args) {
+        try {
+            return method.invoke(o, args);
+        } catch (InvocationTargetException e) {
+            if (e.getTargetException() instanceof WebApplicationException) {
+                return ((WebApplicationException) e.getTargetException()).getResponse();
+            }
+
+            // other exceptions differ based on the participant method type
+            if (type.equals(COMPENSATE)) {
+                return ParticipantStatus.FailedToCompensate.name();
+            } else if (type.equals(COMPLETE)) {
+                return ParticipantStatus.FailedToComplete.name();
+            } else {
+                // @Status and @Forget should return HTTP 500
+                return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+            }
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Response processResult(Object result, URI lraId, Method method, String type) {
+        Response.ResponseBuilder builder = Response.status(Response.Status.OK);
+
+        if (result instanceof CompletionStage) {
+            // store the CompletionStage result and respond compensating / completing
+            participantStatusMap.put(lraId, Optional.empty());
+            ((CompletionStage<?>) result)
+                .thenAccept(res -> participantStatusMap.replace(lraId, Optional.ofNullable(res)))
+                .exceptionally(throwable -> {
+                    participantStatusMap.replace(lraId, Optional.ofNullable(throwable));
```

Comment:
So you correctly remember that the participant threw an exception but you never act upon it. In the `if then else` block below you need:

```java
        } else if (result instanceof Throwable) {
            if type is compensate
                return ParticipantStatus.FailedToCompensate.name();
            else if type is complete
                return ParticipantStatus.FailedToComplete.name();
```

I still think that all of these optional signatures is overly complex for both the application developer and for the spec implementer: doing specs by committee is well known to result in suboptimal results and that is what we have here. I think the mistakes you have made in the last review and in this latest review only serve to highlight these facts. My original solution did not suffer from either of these problems, mainly because the original feature was simply stated. I think that any future potential spec implementors will take one look at this stuff and decide, no, it is too complicated.